activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -203,11 +223,41 @@ def begin_transaction(options = {})
                                        run_commit_callbacks: run_commit_callbacks)
             end
 
+          transaction.materialize! unless @connection.supports_lazy_transactions? && lazy_transactions_enabled?
           @stack.push(transaction)
+          @has_unmaterialized_transactions = true if @connection.supports_lazy_transactions?
           transaction
         end
       end
 
+      def disable_lazy_transactions!
+        materialize_transactions
+        @lazy_transactions_enabled = false
+      end
+
+      def enable_lazy_transactions!
+        @lazy_transactions_enabled = true
+      end
+
+      def lazy_transactions_enabled?
+        @lazy_transactions_enabled
+      end
+
+      def materialize_transactions
+        return if @materializing_transactions
+        return unless @has_unmaterialized_transactions
+
+        @connection.lock.synchronize do
+          begin
+            @materializing_transactions = true
+            @stack.drop_while(&:materialized?).each(&:materialize!)
```

Comment:
Each `drop_while` creates an extra array, so how about adding `return if materialized?` guard in `materialize!`?

Reply:
I'm not sure this is worth optimising. The `return unless @has_unmaterialized_transactions` guard already covers the case where there are no transactions to materialize, so we only allocate an extra array when we're about to execute a query, which will result in many allocations.

The use of `drop_while` also expresses an invariant: `@stack` contains zero or more materialized transactions, followed by zero or more unmaterialized transactions. `@stack.each(&:materialize!)` doesn't say anything about the structure of the data.

If you still think this is worth doing, I'd prefer to check `materialized?` inline, both to avoid duplicating the check across both implementations of `materialize!` and to make it clearer that we're not materializing _all_ transactions:

```ruby
@stack.each { |t| t.materialize! unless t.materialized? }
```

Reply:
I'm convinced - I pushed an update. Thanks for the explanation @kamipo 🙇